src/main/java/org/hydev/mcpm/client/installer/InstallInteractor.java

```diff
@@ -0,0 +1,99 @@
+package org.hydev.mcpm.client.installer;
+
+import org.hydev.mcpm.client.Downloader;
+import org.hydev.mcpm.client.database.DatabaseInteractor;
+import org.hydev.mcpm.client.database.LocalPluginTracker;
+import org.hydev.mcpm.client.database.boundary.SearchPackagesBoundary;
+import org.hydev.mcpm.client.database.fetcher.LocalDatabaseFetcher;
+import org.hydev.mcpm.client.database.inputs.SearchPackagesInput;
+import org.hydev.mcpm.client.database.inputs.SearchPackagesType;
+import org.hydev.mcpm.client.database.results.ListPackagesResult;
+import org.hydev.mcpm.client.database.results.SearchPackagesResult;
+import org.hydev.mcpm.client.injector.PluginNotFoundException;
+import org.hydev.mcpm.client.installer.input.InstallInput;
+import org.hydev.mcpm.client.models.PluginModel;
+import org.hydev.mcpm.client.models.PluginYml;
+
+import java.io.File;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+
+public class InstallInteractor implements InstallBoundary {
+    @Override
+    public void installPlugin(InstallInput installInput) {
+        SearchPackagesInput searchPackagesInput = new SearchPackagesInput(installInput.type(), installInput.name(), false);
+        SearchPackagesResult searchPackageResult = installInput.searchPackage().search(searchPackagesInput);
+        if (searchPackageResult.state() != SearchPackagesResult.State.SUCCESS) {
+            throw new IllegalArgumentException("Illegal");
+        }
+
+        if (searchPackageResult.plugins().isEmpty()) {
+            throw new RuntimeException();
+        }
+
+        PluginModel latestPluginModel = searchPackageResult.plugins().get(0);
+        long id = latestPluginModel.id();
+        for (PluginModel plugin: searchPackageResult.plugins()) {
+            if (plugin.id() > id) {
+                id = plugin.id();
+                latestPluginModel = plugin;
+            }
+        }
+
+        var pluginVersion1 = latestPluginModel.getLatestPluginVersion();
+        if (pluginVersion1.isEmpty()) {
+            throw new RuntimeException();
+        }
+
+        var pluginVersion = pluginVersion1.get();
+
+        /*
+    * 1. Search the name and get a list of plugins
+    * 2. Pick the plugin id
+    * 3. Construct the url
+    * 4. Download it (Downloader)
+    * 5. Add plugin tracker to check install and write
+    * 6. Install the list of dependency
+    *
+    * */
+
+        List<PluginYml> pluginInstalled = installInput.pluginTracker().listInstalled();
+        for (PluginYml pluginYml:pluginInstalled) {
+            if (pluginYml != null && pluginYml.name() != null && pluginYml.name().equals(pluginVersion.meta().name())) {
+                throw new RuntimeException();
```

Comment:
This will error if there are dependencies that are installed locally already.
In the context of the install use case, you'd want to skip over dependencies that are already installed.
In this case, return would be more appropriate than an exception.

Reply:
I read it wrong, it'll error if you're trying to install an already installed plugin, but in that case the exception should tell you that instead of having a blank message

Reply:
I take it back
Every dependency is being recursed on, instead of only those that are not installed, meaning that already installed dependencies will throw an exception. This should be a return instead (recursive base case)

Reply:
Oh yep, this is a bug. Although this can be caught. The best solution here would be to make an `InstallResult` object that has one of those state variables. This will kinda force us to deal with these hard questions (when to throw when not to).

Reply:
I don't see why we should catch only to return an error state when we can directly return the error state.
But also remember that fundamentally, DFS returns once for every visit (you can't just return the recursive call, as this would not continue the other required recursive calls).
So do you want to make a state for "no dependencies already installed" and "some dependencies already installed"?
I think it makes more sense to return a list of all dependencies that are already installed, if we care about those. Otherwise I would keep the method void and just put a return here.